### PR TITLE
add responses prop test

### PR DESCRIPTION
**Pull request by bdemann**

### CODE REVIEW

File: tests/property/ic_api/responses/test/tests.ts
Comment (lastmjs) [inline tests/property/ic_api/responses/test/tests.ts:1]: Make sure to test RejectionCode and maybe reject message when there is no rejection...RejectionCode might be 0 for example

File: tests/property/ic_api/responses/test/test.ts [hunk 1]
```diff
@@ -0,0 +1,13 @@
+import { getCanisterId } from 'azle/dfx';
+import { runTests } from 'azle/test';
+
+import { createActor } from './dfx_generated/caller';
+import { getTests } from './tests';
+
+const callerCanister = createActor(getCanisterId('caller'), {
```
Comment (lastmjs) [inline tests/property/ic_api/responses/test/test.ts:7]: Use getCanisterActor

File: tests/property/ic_api/responses/tsconfig.json
Comment (lastmjs) [inline tests/property/ic_api/responses/tsconfig.json:1]: top-level await

**Review by lastmjs — CHANGES_REQUESTED** [review]

File: tests/property/ic_api/responses/src/caller/index.ts [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    query,
+    reject,
+    rejectCode,
+    rejectMessage,
+    reply,
+    trap,
+    update
+} from 'azle';
+
+const RejectionCodeVariant = IDL.Variant({
```
[the remark this answers is not in the document]
Comment (bdemann) [reply tests/property/ic_api/responses/src/caller/index.ts:13]: done

File: tests/property/ic_api/responses/test/tests.ts [hunk 3]
```diff
@@ -0,0 +1,73 @@
+import { ActorSubclass } from '@dfinity/agent';
+import { defaultPropTestParams, expect, it, Test } from 'azle/test';
+import fc from 'fast-check';
+
+// @ts-ignore this path may not exist when these tests are imported into other test projects
```
[the remark this answers is not in the document]
Comment (bdemann) [reply tests/property/ic_api/responses/test/tests.ts:5]: done

File: tests/property/ic_api/responses/tsconfig.json
Comment (bdemann) [reply tests/property/ic_api/responses/tsconfig.json:1]: Can't because of jest

File: tests/property/ic_api/responses/test/test.ts (hunk 1 above)
Comment (bdemann) [reply tests/property/ic_api/responses/test/test.ts:7]: done

File: tests/property/ic_api/responses/src/caller/index.ts [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    @update([IDL.Text], IDL.Text)
+    async echoThroughReject(message: string): Promise<string> {
+        try {
+            await call(getRejectorPrincipal(), 'echoReject', {
+                paramIdlTypes: [IDL.Text],
+                returnIdlType: IDL.Text,
+                args: [message]
+            });
+        } catch (error) {
+            console.log('echoThroughReject caught an error', error);
```
[the remark this answers is not in the document]
Comment (bdemann) [reply tests/property/ic_api/responses/src/caller/index.ts:75]: done

File: tests/property/ic_api/responses/src/caller/index.ts [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+            });
+        } catch (error) {
+            console.log('getRejectCodeCanisterReject caught an error', error);
+            return rejectCode();
+        }
+        throw new Error('This should never be thrown');
+    }
+}
+
+function getRejectorPrincipal(): string {
```
[the remark this answers is not in the document]
Comment (bdemann) [reply tests/property/ic_api/responses/src/caller/index.ts:108]: done

File: tests/property/ic_api/responses/src/caller/index.ts [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+            });
+        } catch (error) {
+            console.log('echoThroughReject caught an error', error);
+            return rejectMessage();
+        }
+        throw new Error('This should never be thrown');
+    }
+
+    @update([], RejectionCodeVariant)
+    async getRejectCodeCanisterError(): Promise<RejectionCodeVariant> {
```
[the remark this answers is not in the document]
Comment (bdemann) [reply tests/property/ic_api/responses/src/caller/index.ts:82]: done

File: tests/property/ic_api/responses/test/tests.ts
Comment (bdemann) [reply tests/property/ic_api/responses/test/tests.ts:1]: done

File: tests/property/ic_api/responses/src/rejector/index.ts [hunk 7]
```diff
@@ -0,0 +1,18 @@
+import { IDL, query, reject } from 'azle';
+
+export default class {
+    @query([IDL.Text], IDL.Text, { manual: true })
```
[the remark this answers is not in the document]
Comment (bdemann) [reply tests/property/ic_api/responses/src/rejector/index.ts:4]: Either it needs to be manual, or I'm doing it wrong, or azle is broken

File: tests/property/ic_api/responses/src/caller/index.ts
[the remark this answers is not in the document]
Comment (bdemann) [reply tests/property/ic_api/responses/src/caller/index.ts:1]: I don't think so

File: tests/property/ic_api/responses/src/caller/index.ts [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+          CanisterError: null;
+      }
+    | {
+          Unknown: null;
+      };
+
+export default class {
+    @query([IDL.Text], IDL.Text, { manual: true })
+    alwaysReplyQuery(input: string): void {
+        reply({ data: input, idlType: IDL.Text });
```
[the remark this answers is not in the document]
Comment (bdemann) [reply tests/property/ic_api/responses/src/caller/index.ts]: done

File: tests/property/ic_api/responses/src/caller/index.ts [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+
+    @query([IDL.Text], IDL.Text, { manual: true })
+    alwaysRejectQuery(message: string): void {
+        reject(message);
+    }
+
+    @query([IDL.Int], IDL.Int, { manual: true })
+    evenOrRejectQuery(number: bigint): void {
+        if (number % 2n === 0n) {
+            reply({ data: number, idlType: IDL.Int });
```
[the remark this answers is not in the document]
Comment (bdemann) [reply tests/property/ic_api/responses/src/caller/index.ts]: I don't think this one needs to be arbitrary. We are just testing to make sure that it's not always reply or always reject, the ability to do arbitrary values is sufficiently tests elsewhere

File: tests/property/ic_api/responses/test/tests.ts [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+import { _SERVICE } from './dfx_generated/caller/caller.did';
+
+export function getTests(callerCanister: ActorSubclass<_SERVICE>): Test {
+    return () => {
+        it('should always reply with the input in alwaysReplyQuery', async () => {
+            await fc.assert(
+                fc.asyncProperty(fc.string(), async (input) => {
```
[the remark this answers is not in the document]
Comment (bdemann) [reply tests/property/ic_api/responses/test/tests.ts]: done

File: tests/property/ic_api/responses/test/tests.ts [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+                fc.asyncProperty(fc.string(), async (message) => {
+                    await expect(
+                        callerCanister.alwaysRejectQuery(message)
+                    ).rejects.toThrow(escapeCandidString(message));
+                }),
+                defaultPropTestParams
+            );
+        });
+
+        it('should reply with even numbers and reject odd numbers in evenOrRejectQuery', async () => {
```
[the remark this answers is not in the document]
Comment (bdemann) [reply tests/property/ic_api/responses/test/tests.ts]: I don't think we need it here

File: tests/property/ic_api/responses/src/caller/index.ts
[the remark this answers is not in the document]
Comment (lastmjs) [reply tests/property/ic_api/responses/src/caller/index.ts:1]: Can you test this? I think you can. I think I just did it a few days ago.

File: tests/property/ic_api/rejectCode/src/rejector/index.ts [hunk 12]
```diff
@@ -15,4 +10,9 @@ export default class {
     rejectWithMessage(): void {
         reject('Rejection with message');
     }
+
+    @query([], undefined, { manual: true })
```
Comment (lastmjs) [inline tests/property/ic_api/rejectCode/src/rejector/index.ts:14]: Why is this manual and why is it using reply?

File: tests/property/ic_api/rejectMessage/src/caller/index.ts
Comment (lastmjs) [inline tests/property/ic_api/rejectMessage/src/caller/index.ts:1]: I feel like we should test reject callback and no callback for all of these, let's discuss

File: tests/property/ic_api/rejectMessage/test/tests.ts
Comment (lastmjs) [inline tests/property/ic_api/rejectMessage/test/tests.ts:1]: This test seems too simple. I would think we should do the following: query and update method with just a reject message, then the same thing through a cross-canister call

File: tests/property/ic_api/reply/test/generate_canister.ts
Comment (lastmjs) [inline tests/property/ic_api/reply/test/generate_canister.ts:1]: I think we should test query and update

File: tests/property/ic_api/reply/test/tests.ts [hunk 13]
```diff
@@ -0,0 +1,86 @@
+globalThis._azleExperimental = true;
+import { ActorSubclass } from '@dfinity/agent';
```
Comment (lastmjs) [inline tests/property/ic_api/reply/test/tests.ts:2]: Line break above please

**Review by lastmjs — CHANGES_REQUESTED** [review]

File: tests/property/ic_api/responses/src/caller/index.ts
[the remark this answers is not in the document]
Comment (bdemann) [reply tests/property/ic_api/responses/src/caller/index.ts:1]: I did test it, I think you can't, I'll show you, maybe I missed something

File: tests/property/ic_api/reject/src/index.ts [hunk 14]
```diff
@@ -0,0 +1,17 @@
+import { IDL, query, reject, reply } from 'azle';
+
+export default class {
+    @query([IDL.Text], IDL.Text, { manual: true })
```
[the remark this answers is not in the document]
Comment (bdemann) [reply tests/property/ic_api/reject/src/index.ts:4]: I still think you do

File: tests/property/ic_api/reject/src/index.ts [hunk 15]
```diff
@@ -0,0 +1,17 @@
...
+    alwaysRejectQuery(message: string): void {
+        reject(message);
+    }
+
+    @query([IDL.Int], IDL.Int, { manual: true })
+    evenOrRejectQuery(number: bigint): void {
```
[the remark this answers is not in the document]
Comment (bdemann) [reply tests/property/ic_api/reject/src/index.ts:10]: okay

File: tests/property/ic_api/reject/src/index.ts [hunk 16]
```diff
@@ -0,0 +1,17 @@
+import { IDL, query, reject, reply } from 'azle';
+
+export default class {
+    @query([IDL.Text], IDL.Text, { manual: true })
+    alwaysRejectQuery(message: string): void {
+        reject(message);
+    }
```
[the remark this answers is not in the document]
Comment (bdemann) [reply tests/property/ic_api/reject/src/index.ts:7]: done

File: tests/property/ic_api/reject/test/tests.ts [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+    it,
+    Test
+} from 'azle/test';
+import fc from 'fast-check';
+
+import { _SERVICE as Actor } from './dfx_generated/canister/canister.did';
+
+export function getTests(): Test {
+    return () => {
+        it('should always reject with the provided message in alwaysRejectQuery', async () => {
```
[the remark this answers is not in the document]
Comment (bdemann) [reply tests/property/ic_api/reject/test/tests.ts:14]: Well... I mean assuming that not manual works

File: tests/property/ic_api/rejectCode/.gitignore
[the remark this answers is not in the document]
Comment (bdemann) [reply tests/property/ic_api/rejectCode/.gitignore:1]: done

File: tests/property/ic_api/rejectMessage/src/caller/index.ts
[the remark this answers is not in the document]
Comment (bdemann) [reply tests/property/ic_api/rejectMessage/src/caller/index.ts:1]: done

File: tests/property/ic_api/reply/test/generate_canister.ts
Comment (bdemann) [reply tests/property/ic_api/reply/test/generate_canister.ts:1]: done

File: tests/property/ic_api/reply/test/tests.ts (hunk 13 above)
Comment (bdemann) [reply tests/property/ic_api/reply/test/tests.ts:2]: done

File: tests/property/ic_api/reject/src/index.ts [hunk 18]
```diff
@@ -1,11 +1,16 @@
-import { IDL, query, reject, reply } from 'azle';
+import { IDL, query, reject, reply, update } from 'azle';
 
 export default class {
     @query([IDL.Text], IDL.Text, { manual: true })
     alwaysRejectQuery(message: string): void {
         reject(message);
     }
 
+    @update([IDL.Text], IDL.Text, { manual: true })
```
Comment (lastmjs) [inline tests/property/ic_api/reject/src/index.ts:9]: Where are the tests for this?

**Review by lastmjs — CHANGES_REQUESTED** [review]

File: tests/property/ic_api/rejectCode/src/rejector/index.ts (hunk 12 above)
Comment (bdemann) [reply tests/property/ic_api/rejectCode/src/rejector/index.ts]: done

**Review by lastmjs — APPROVED** [review]

**Comment by bdemann** [comment]
This is blocked by a weird issue where on the second run the actor that it has is the actor from the previous run

**Review by lastmjs — APPROVED** [review]

**Pull request merged**